lib/iris/tests/unit/constraints/test_Constraint_equality.py

```diff
@@ -0,0 +1,266 @@
+        c1 = AttributeConstraint(a=lambda a: False)
+        c2 = AttributeConstraint(a=lambda a: False)
+        self.assertNotEqual(c1, c2)
+
+
+class Test_NameConstraint__hash__(tests.IrisTest):
+    def test_empty(self):
+        c1 = NameConstraint()
+        c2 = NameConstraint()
+        self.assertEqual(hash(c1), hash(c1))
+        self.assertNotEqual(hash(c1), hash(c2))
+
+
+class Test_NameConstraint__eq__(tests.IrisTest):
+    def test_empty_same(self):
+        c1 = NameConstraint()
+        c2 = NameConstraint()
+        self.assertEqual(c1, c2)
+        self.assertIsNot(c1, c2)
+
+    def test_attribute_plain_empty_diff(self):
+        c1 = NameConstraint()
+        c2 = Constraint()
+        self.assertNotEqual(c1, c2)
+
+    def test_names_same(self):
+        c1 = NameConstraint(standard_name="air_temperature")
+        c2 = NameConstraint(standard_name="air_temperature")
+        self.assertEqual(c1, c2)
+
+    def test_full_same(self):
+        c1 = NameConstraint(
+            standard_name="air_temperature",
+            long_name="temp",
+            var_name="tair",
+            STASH="m01s02i003",
+        )
+        c2 = NameConstraint(
+            standard_name="air_temperature",
+            long_name="temp",
+            var_name="tair",
+            STASH="m01s02i003",
+        )
+        self.assertEqual(c1, c2)
+
+    def test_missing_diff(self):
+        c1 = NameConstraint(standard_name="air_temperature", var_name="tair")
+        c2 = NameConstraint(standard_name="air_temperature")
+        self.assertNotEqual(c1, c2)
+
+    def test_standard_name_diff(self):
+        c1 = NameConstraint(standard_name="air_temperature")
+        c2 = NameConstraint(standard_name="height")
+        self.assertNotEqual(c1, c2)
+
+    def test_long_name_diff(self):
+        c1 = NameConstraint(long_name="temp")
+        c2 = NameConstraint(long_name="t3")
+        self.assertNotEqual(c1, c2)
+
+    def test_var_name_diff(self):
+        c1 = NameConstraint(var_name="tair")
+        c2 = NameConstraint(var_name="xxx")
+        self.assertNotEqual(c1, c2)
+
+    def test_stash_diff(self):
+        c1 = NameConstraint(STASH="m01s02i003")
+        c2 = NameConstraint(STASH="m01s02i777")
+        self.assertNotEqual(c1, c2)
+
+    def test_func_same(self):
+        def func(name):
+            return True
+
+        c1 = NameConstraint(STASH="m01s02i003", long_name=func)
+        c2 = NameConstraint(STASH="m01s02i003", long_name=func)
+        self.assertEqual(c1, c2)
+
+    def test_func_diff(self):
+        c1 = NameConstraint(STASH="m01s02i003", long_name=lambda n: True)
+        c2 = NameConstraint(STASH="m01s02i003", long_name=lambda n: True)
+        self.assertNotEqual(c1, c2)
+
+
+class Test___hash__(tests.IrisTest):
+    def test_empty(self):
+        c1 = Constraint() & Constraint()
+        c2 = Constraint() & Constraint()
+        self.assertEqual(hash(c1), hash(c1))
+        self.assertNotEqual(hash(c1), hash(c2))
```

Comment:
I think it would also be worth testing the behaviour of two different combinations of an identical pair of constraints. i.e.
```python
c1 = Constraint()
c2 = Constraint()
cc1 = c1 & c2
cc2 = c1 & c2
```

Reply:
:+1: